packages/utils/src/lib/throttle.ts

```diff
@@ -5,17 +5,126 @@ const isTest = () =>
+let windowStartTime = 0
+let maxDetectedFps = DEFAULT_FPS // Track the highest FPS we've measured
+let isManuallySet = false // Track if user has manually set the FPS
+
+const updateTargetFps = () => {
+	if (fpsCheckHistory.length < 3) return // Need at least 3 measurements
+
+	const avgFps = fpsCheckHistory.reduce((a, b) => a + b, 0) / fpsCheckHistory.length
+
+	// Track the maximum FPS we've detected (likely the device's natural refresh rate)
+	if (avgFps > maxDetectedFps) {
+		maxDetectedFps = Math.min(MAX_FPS, Math.round(avgFps))
+
+		// Only update target FPS if user hasn't manually set it
+		if (!isManuallySet) {
+			targetFps = maxDetectedFps
+			targetTimePerFrame = getTargetTimePerFrame(targetFps)
+		}
+	}
+
+	// Keep only recent measurements
+	if (fpsCheckHistory.length > 5) {
+		fpsCheckHistory = fpsCheckHistory.slice(-3)
+	}
+}
+
+// Track actual browser frame timing using requestAnimationFrame
+let frameTimingRaf: number | undefined
+let isFrameTimingActive = false
+let measurementStartTime = 0
+const MEASUREMENT_DURATION = 2000 // Continue measuring for 2 seconds after throttling activity
+
+const startFrameTimingMeasurement = () => {
+	if (isFrameTimingActive || isTest()) return
+
+	isFrameTimingActive = true
+	framesInCurrentWindow = 0
+	windowStartTime = performance.now()
+	measurementStartTime = windowStartTime
+
+	const measureFrame = () => {
+		// Count this frame
+		framesInCurrentWindow++
+
+		// Calculate elapsed time
+		const now = performance.now()
+		const elapsed = now - windowStartTime
+
+		// Check if we should calculate FPS
+		if (elapsed >= FPS_MEASUREMENT_WINDOW) {
+			const measuredFps = Math.round(
+				framesInCurrentWindow * (FPS_MEASUREMENT_WINDOW / elapsed) * (1000 / FPS_MEASUREMENT_WINDOW)
+			)
+
+			// Store the measurement
+			fpsCheckHistory.push(measuredFps)
+
+			// Reset for next measurement window
+			framesInCurrentWindow = 0
+			windowStartTime = now
+
+			// Check if we should update target FPS
+			if (fpsCheckHistory.length >= 3) {
+				updateTargetFps()
+			}
+		}
+
+		// Continue measuring for a reasonable period even if queue is empty
+		// This ensures we collect enough data for adaptive FPS decisions
+		const timeSinceStart = now - measurementStartTime
+		if (fpsQueue.length > 0 || timeSinceStart < MEASUREMENT_DURATION) {
+			// eslint-disable-next-line no-restricted-globals
+			frameTimingRaf = requestAnimationFrame(measureFrame)
+		} else {
+			isFrameTimingActive = false
+		}
+	}
+
+	// eslint-disable-next-line no-restricted-globals
+	frameTimingRaf = requestAnimationFrame(measureFrame)
+}
+
+const stopFrameTimingMeasurement = () => {
+	if (frameTimingRaf) {
+		cancelAnimationFrame(frameTimingRaf)
+		frameTimingRaf = undefined
+	}
+	isFrameTimingActive = false
+}
+
 const flush = () => {
 	const queue = fpsQueue.splice(0, fpsQueue.length)
 	for (const fn of queue) {
 		fn()
 	}
+
+	// Start measuring actual browser frame timing
+	if (!isTest()) {
+		startFrameTimingMeasurement()
```

Comment:
Should we maybe do this outside of flush? This is a really hot code path so might be worth it to avoid this call which early returns all the time?